File: typr-tester-oracle-kotlin/src/kotlin/oracledb/all_scalar_types/AllScalarTypesTest.kt
package oracledb.all_scalar_types

import oracledb.OracleTestHelper
import oracledb.customtypes.Defaulted
import org.junit.Assert.*
import org.junit.Test
import java.math.BigDecimal
import java.time.LocalDateTime

class AllScalarTypesTest {
    private val repo = AllScalarTypesRepoImpl()

    @Test
    fun testInsertAndSelectAllScalarTypes() {
        OracleTestHelper.run { c ->
            val unsaved = AllScalarTypesRowUnsaved(
                "test varchar2",
                BigDecimal("123.45"),
                LocalDateTime.of(2025, 1, 1, 12, 0),
                LocalDateTime.of(2025, 1, 1, 12, 0, 30, 123456000),
                "test clob content",
                "not null value",
                Defaulted.UseDefault()
            )

            val insertedRow = repo.insert(unsaved, c)

            assertNotNull(insertedRow)
            val inserted = repo.selectById(insertedRow.id, c)
            assertNotNull(inserted)
            assertEquals("test varchar2", inserted!!.colVarchar2)
            assertEquals(BigDecimal("123.45"), inserted.colNumber)
            assertNotNull(inserted.colDate)
            assertNotNull(inserted.colTimestamp)
            assertEquals("test clob content", inserted.colClob)
            assertEquals("not null value", inserted.colNotNull)
        }
    }

    @Test
    fun testInsertWithNullOptionalColumns() {
        OracleTestHelper.run { c ->
            val unsaved = AllScalarTypesRowUnsaved(
                null,
                null,
                null,
                null,
                null,
                "required value",
                Defaulted.UseDefault()
            )

            val insertedRow = repo.insert(unsaved, c)
            val inserted = repo.selectById(insertedRow.id, c)!!

            assertNull(inserted.colVarchar2)
            assertNull(inserted.colNumber)
            assertNull(inserted.colDate)
            assertNull(inserted.colTimestamp)
            assertNull(inserted.colClob)
            assertEquals("required value", inserted.colNotNull)
        }
    }

    @Test
    fun testUpdateScalarTypes() {
        OracleTestHelper.run { c ->
            val unsaved = AllScalarTypesRowUnsaved(
                "original",
                BigDecimal("100.00"),
                null,
                null,
                null,
                "not null",
                Defaulted.UseDefault()
            )

            val insertedRow = repo.insert(unsaved, c)
            val inserted = repo.selectById(insertedRow.id, c)!!

            val updated = inserted.copy(
                colVarchar2 = "updated",
                colNumber = BigDecimal("200.01")
            )

            val wasUpdated = repo.update(updated, c)
            assertTrue(wasUpdated)

            val fetched = repo.selectById(insertedRow.id, c)!!
            assertEquals("updated", fetched.colVarchar2)
            assertEquals(BigDecimal("200.01"), fetched.colNumber)
        }
    }

    @Test
    fun testDeleteScalarTypes() {
        OracleTestHelper.run { c ->
            val unsaved = AllScalarTypesRowUnsaved(
                "to delete",
                null,
                null,
                null,
                null,
                "delete me",
                Defaulted.UseDefault()
            )

            val insertedRow = repo.insert(unsaved, c)

            val deleted = repo.deleteById(insertedRow.id, c)
            assertTrue(deleted)

            val found = repo.selectById(insertedRow.id, c)
            assertNull(found)
        }
    }

    @Test
    fun testSelectAll() {
        OracleTestHelper.run { c ->
            repo.insert(AllScalarTypesRowUnsaved("row1", null, null, null, null, "not null 1", Defaulted.UseDefault()), c)
            repo.insert(AllScalarTypesRowUnsaved("row2", null, null, null, null, "not null 2", Defaulted.UseDefault()), c)

            val all = repo.selectAll(c)
            assertTrue(all.size >= 2)
        }
    }

    @Test
    fun testScalarTypeVarchar2() {
        OracleTestHelper.run { c ->
            val longString = "A".repeat(100)
            val unsaved = AllScalarTypesRowUnsaved(
                longString,
                null,
                null,
                null,
                null,
                "required",
                Defaulted.UseDefault()
            )

            val insertedRow = repo.insert(unsaved, c)
            assertEquals(longString, insertedRow.colVarchar2)
        }
    }

    @Test
    fun testScalarTypeNumber() {
        OracleTestHelper.run { c ->
            val unsaved = AllScalarTypesRowUnsaved(
                null,
                BigDecimal("999999.99"),
                null,
                null,
                null,
                "required",
                Defaulted.UseDefault()
            )

            val insertedRow = repo.insert(unsaved, c)
            assertEquals(BigDecimal("999999.99"), insertedRow.colNumber)
        }
    }

    @Test
    fun testBigDecimalPrecision() {
        OracleTestHelper.run { c ->
            // Column is NUMBER(10,2) - max 8 digits before decimal, 2 after
            val preciseNumber = BigDecimal("12345678.99")
            val unsaved = AllScalarTypesRowUnsaved(
                null,
                preciseNumber,
                null,
                null,
                null,
                "precision test",
                Defaulted.UseDefault()
            )

            val insertedRow = repo.insert(unsaved, c)
            val inserted = repo.selectById(insertedRow.id, c)!!

            assertNotNull(inserted.colNumber)
            assertEquals(preciseNumber, inserted.colNumber)
        }
    }

    @Test
    fun testDateTimeColumns() {
        OracleTestHelper.run { c ->
            val dateValue = LocalDateTime.of(2025, 6, 15, 10, 30, 0)
            val timestampValue = LocalDateTime.of(2025, 6, 15, 10, 30, 45, 123456789)

            val unsaved = AllScalarTypesRowUnsaved(
                null,
                null,
                dateValue,
                timestampValue,
                null,
                "datetime test",
                Defaulted.UseDefault()
            )

            val insertedRow = repo.insert(unsaved, c)
            val inserted = repo.selectById(insertedRow.id, c)!!

            // DATE in Oracle includes time down to seconds
            assertNotNull(inserted.colDate)
            assertEquals(2025, inserted.colDate!!.year)
            assertEquals(6, inserted.colDate!!.monthValue)
            assertEquals(15, inserted.colDate!!.dayOfMonth)

            // TIMESTAMP includes fractional seconds
            assertNotNull(inserted.colTimestamp)
        }
    }

    @Test
    fun testClobColumn() {
        OracleTestHelper.run { c ->
            val largeText = "A".repeat(10000)
            val unsaved = AllScalarTypesRowUnsaved(
                null,
                null,
                null,
                null,
                largeText,
                "clob test",
                Defaulted.UseDefault()
            )

            val insertedRow = repo.insert(unsaved, c)
            val inserted = repo.selectById(insertedRow.id, c)!!

            assertEquals(largeText, inserted.colClob)
        }
    }
}

File: typr-tester-oracle-scala-new/src/scala/oracledb/all_scalar_types/AllScalarTypesTest.scala
package oracledb.all_scalar_types

import oracledb.customtypes.Defaulted
import oracledb.withConnection
import org.scalatest.funsuite.AnyFunSuite

import java.time.LocalDateTime

class AllScalarTypesTest extends AnyFunSuite {
  val repo: AllScalarTypesRepoImpl = new AllScalarTypesRepoImpl

  test("insert and select all scalar types") {
    withConnection { c =>
      given java.sql.Connection = c
      val unsaved = new AllScalarTypesRowUnsaved(
        Some("test varchar2"),
        Some(BigDecimal("123.45")),
        Some(LocalDateTime.of(2025, 1, 1, 12, 0)),
        Some(LocalDateTime.of(2025, 1, 2, 13, 30)),
        Some("test clob content"),
        "required not null field",
        Defaulted.UseDefault[AllScalarTypesId]()
      )

      val inserted = repo.insert(unsaved)

      val _ = assert(inserted.colVarchar2 == Some("test varchar2"))
      val _ = assert(inserted.colNumber == Some(BigDecimal("123.45")))
      val _ = assert(inserted.colDate.isDefined)
      val _ = assert(inserted.colTimestamp.isDefined)
      val _ = assert(inserted.colClob == Some("test clob content"))
      val _ = assert(inserted.colNotNull == "required not null field")

      val found = repo.selectById(inserted.id)
      val _ = assert(found.isDefined)
      val _ = assert(inserted == found.get)
    }
  }

  test("insert with null values") {
    withConnection { c =>
      given java.sql.Connection = c
      val unsaved = new AllScalarTypesRowUnsaved(
        None,
        None,
        None,
        None,
        None,
        "only required field",
        Defaulted.UseDefault[AllScalarTypesId]()
      )

      val inserted = repo.insert(unsaved)

      val _ = assert(inserted.colVarchar2.isEmpty)
      val _ = assert(inserted.colNumber.isEmpty)
      val _ = assert(inserted.colDate.isEmpty)
      val _ = assert(inserted.colTimestamp.isEmpty)
      val _ = assert(inserted.colClob.isEmpty)
      val _ = assert(inserted.colNotNull == "only required field")
    }
  }

  test("update all scalar types") {
    withConnection { c =>
      given java.sql.Connection = c
      val unsaved = new AllScalarTypesRowUnsaved(
        Some("original"),
        Some(BigDecimal("100.00")),
        None,
        None,
        None,
        "required",
        Defaulted.UseDefault[AllScalarTypesId]()
      )

      val inserted = repo.insert(unsaved)

      val updatedRow = inserted.copy(
        colVarchar2 = Some("updated"),
        colNumber = Some(BigDecimal("200.01"))
      )

      val wasUpdated = repo.update(updatedRow)
      val _ = assert(wasUpdated)

      val fetched = repo.selectById(inserted.id).get
      val _ = assert(fetched.colVarchar2 == Some("updated"))
      val _ = assert(fetched.colNumber == Some(BigDecimal("200.01")))
    }
  }

  test("delete all scalar types") {
    withConnection { c =>
      given java.sql.Connection = c
      val unsaved = new AllScalarTypesRowUnsaved(
        Some("to delete"),
        None,
        None,
        None,
        None,
        "required",
        Defaulted.UseDefault[AllScalarTypesId]()
      )

      val inserted = repo.insert(unsaved)

      val deleted = repo.deleteById(inserted.id)
      val _ = assert(deleted)

      val found = repo.selectById(inserted.id)
      val _ = assert(found.isEmpty)
    }
  }

  test("select all") {
    withConnection { c =>
      given java.sql.Connection = c
      val unsaved1 = new AllScalarTypesRowUnsaved(
        Some("row 1"),
        None,
        None,
        None,
        None,
        "required 1",
        Defaulted.UseDefault[AllScalarTypesId]()
      )

      val unsaved2 = new AllScalarTypesRowUnsaved(
        Some("row 2"),
        None,
        None,
        None,
        None,
        "required 2",
        Defaulted.UseDefault[AllScalarTypesId]()
      )

      val _ = repo.insert(unsaved1)
      val _ = repo.insert(unsaved2)

      val all = repo.selectAll
      val _ = assert(all.length >= 2)
    }
  }

  test("scalar type varchar2") {
    withConnection { c =>
      given java.sql.Connection = c
      val longString = "A" * 100
      val unsaved = new AllScalarTypesRowUnsaved(
        Some(longString),
        None,
        None,
        None,
        None,
        "required",
        Defaulted.UseDefault[AllScalarTypesId]()
      )

      val inserted = repo.insert(unsaved)
      val _ = assert(inserted.colVarchar2 == Some(longString))
    }
  }

  test("scalar type number") {
    withConnection { c =>
      given java.sql.Connection = c
      val unsaved = new AllScalarTypesRowUnsaved(
        None,
        Some(BigDecimal("999999.99")),
        None,
        None,
        None,
        "required",
        Defaulted.UseDefault[AllScalarTypesId]()
      )

      val inserted = repo.insert(unsaved)
      val _ = assert(inserted.colNumber == Some(BigDecimal("999999.99")))
    }
  }

  test("scalar type date") {
    withConnection { c =>
      given java.sql.Connection = c
      val testDate = LocalDateTime.of(2025, 12, 31, 23, 59, 59)
      val unsaved = new AllScalarTypesRowUnsaved(
        None,
        None,
        Some(testDate),
        None,
        None,
        "required",
        Defaulted.UseDefault[AllScalarTypesId]()
      )

      val inserted = repo.insert(unsaved)
      val _ = assert(inserted.colDate.isDefined)
      val _ = assert(inserted.colDate.get.getYear == 2025)
      val _ = assert(inserted.colDate.get.getMonthValue == 12)
      val _ = assert(inserted.colDate.get.getDayOfMonth == 31)
    }
  }

  test("scalar type timestamp") {
    withConnection { c =>
      given java.sql.Connection = c
      val testTimestamp = LocalDateTime.of(2025, 6, 15, 14, 30, 45)
      val unsaved = new AllScalarTypesRowUnsaved(
        None,
        None,
        None,
        Some(testTimestamp),
        None,
        "required",
        Defaulted.UseDefault[AllScalarTypesId]()
      )

      val inserted = repo.insert(unsaved)
      val _ = assert(inserted.colTimestamp.isDefined)
      val _ = assert(inserted.colTimestamp.get.getYear == testTimestamp.getYear)
      val _ = assert(inserted.colTimestamp.get.getMonthValue == testTimestamp.getMonthValue)
      val _ = assert(inserted.colTimestamp.get.getDayOfMonth == testTimestamp.getDayOfMonth)
    }
  }

  test("scalar type clob") {
    withConnection { c =>
      given java.sql.Connection = c
      val largeClobContent = "Clob content " * 1000
      val unsaved = new AllScalarTypesRowUnsaved(
        None,
        None,
        None,
        None,
        Some(largeClobContent),
        "required",
        Defaulted.UseDefault[AllScalarTypesId]()
      )

      val inserted = repo.insert(unsaved)
      val _ = assert(inserted.colClob == Some(largeClobContent))
    }
  }
}
